### PR TITLE
fix(npm): support frozen-lockfile in .yarnrc

### LINKED ISSUE #11356: Renovate doesn't update yarn.lock anymore with `--install.frozen-lockfile true` in .yarnrc
### How are you running Renovate?

Self-hosted

### Please select which platform you are using if self-hosting.

GitLab self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

v26.3.0

### Describe the bug

Renovate doesn't update the `yarn.lock` file anymore if the option `--install.frozen-lockfile true` is present in the `.yarnrc` file.

I think I already found the reason for that, in [this commit](https://github.com/renovatebot/renovate/commit/2fe494616da0231278913c55ace8f879c04b9ef4#r55190824).

I created a [minimal reproduction repository](https://github.com/adnotam/renovate_test), but currently is not failing because it's still using the v25.76.2.

Many thanks for your support.

### Relevant debug logs

<details><summary>Logs</summary>

```
DEBUG: Getting updated lock files (repository=ad-notam/website, branch=renovate/webpack-packages)
DEBUG: Writing package.json files (repository=ad-notam/website, branch=renovate/webpack-packages)
       "packageFiles": ["package.json"]
DEBUG: Writing any updated package files (repository=ad-notam/website, branch=renovate/webpack-packages)
DEBUG: Writing package.json (repository=ad-notam/website, branch=renovate/webpack-packages)
DEBUG: No npmrc file found in repository (repository=ad-notam/website, branch=renovate/webpack-packages)
DEBUG: Writing updated .npmrc file to /tmp/renovate/repos/gitlab/ad-notam/website/.npmrc (repository=ad-notam/website, branch=renovate/webpack-packages)
DEBUG: Generating yarn.lock for . (repository=ad-notam/website, branch=renovate/webpack-packages)
DEBUG: Spawning yarn install to create /tmp/renovate/repos/gitlab/ad-notam/website/yarn.lock (repository=ad-notam/website, branch=renovate/webpack-packages)
DEBUG: Updating yarn.lock only - skipping node_modules (repository=ad-notam/website, branch=renovate/webpack-packages)
DEBUG: No node constraint found - using latest (repository=ad-notam/website, branch=renovate/webpack-packages)
DEBUG: Using docker to execute (repository=ad-notam/website, branch=renovate/webpack-packages)
DEBUG: No tag or tagConstraint specified (repository=ad-notam/website, branch=renovate/webpack-packages)
       "image": "docker.io/renovate/node"
DEBUG: Fetching Docker image: docker.io/renovate/node (repository=ad-notam/website, branch=renovate/webpack-packages)
DEBUG: Finished fetching Docker image (repository=ad-notam/website, branch=renovate/webpack-packages)
DEBUG: Executing command (repository=ad-notam/website, branch=renovate/webpack-packages)
       "command": [
         "docker run --rm --name=renovate_node --label=renovate_child -v \"/tmp/renovate/repos/gitlab/ad-notam/website\":\"/tmp/renovate/repos/gitlab/ad-notam/website\" -v \"/tmp/renovate/cache\":\"/tmp/renovate/cache\" -e NPM_CONFIG_CACHE -e npm_config_store -e CI -e YARN_CACHE_FOLDER -w \"/tmp/renovate/repos/gitlab/ad-notam/website\" docker.io/renovate/node bash -l -c \"npm i -g yarn && sed -i 's/ steps,/ steps.slice(0,1),/' /home/ubuntu/.npm-global/lib/node_modules/yarn/lib/cli.js && yarn install --ignore-engines --ignore-platform --network-timeout 100000 --ignore-scripts\""
       ]
DEBUG: rawExec err (repository=ad-notam/website, branch=renovate/webpack-packages)
       "err": {
         "killed": false,
         "code": 1,
         "signal": null,
         "cmd": "docker run --rm --name=renovate_node --label=renovate_child -v \"/tmp/renovate/repos/gitlab/ad-notam/website\":\"/tmp/renovate/repos/gitlab/ad-notam/website\" -v \"/tmp/renovate/cache\":\"/tmp/renovate/cache\" -e NPM_CONFIG_CACHE -e npm_config_store -e CI -e YARN_CACHE_FOLDER -w \"/tmp/renovate/repos/gitlab/ad-notam/website\" docker.io/renovate/node bash -l -c \"npm i -g yarn && sed -i 's/ steps,/ steps.slice(0,1),/' /home/ubuntu/.npm-global/lib/node_modules/yarn/lib/cli.js && yarn install --ignore-engines --ignore-platform --network-timeout 100000 --ignore-scripts\"",
         "stdout": "\n> yarn@1.22.11 preinstall /home/ubuntu/.npm-global/lib/node_modules/yarn\n> :; (node ./preinstall.js > /dev/null 2>&1 || true)\n\n/home/ubuntu/.npm-global/bin/yarn -> /home/ubuntu/.npm-global/lib/node_modules/yarn/bin/yarn.js\n/home/ubuntu/.npm-global/bin/yarnpkg -> /home/ubuntu/.npm-global/lib/node_modules/yarn/bin/yarn.js\n+ yarn@1.22.11\nadded 1 package in 1.013s\nyarn install v1.22.11\n[1/4] Resolving packages...\ninfo Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.\n",
         "stderr": "error Your lockfile needs to be updated, but yarn was run with `--frozen-lockfile`.\n",
         "message": "Command failed: docker run --rm --name=renovate_node --label=renovate_child -v \"/tmp/renovate/repos/gitlab/ad-notam/website\":\"/tmp/renovate/repos/gitlab/ad-notam/website\" -v \"/tmp/renovate/cache\":\"/tmp/renovate/cache\" -e NPM_CONFIG_CACHE -e npm_config_store -e CI -e YARN_CACHE_FOLDER -w \"/tmp/renovate/repos/gitlab/ad-notam/website\" docker.io/renovate/node bash -l -c \"npm i -g yarn && sed -i 's/ steps,/ steps.slice(0,1),/' /home/ubuntu/.npm-global/lib/node_modules/yarn/lib/cli.js && yarn install --ignore-engines --ignore-platform --network-timeout 100000 --ignore-scripts\"\nerror Your lockfile needs to be updated, but yarn was run with `--frozen-lockfile`.\n",
         "stack": "Error: Command failed: docker run --rm --name=renovate_node --label=renovate_child -v \"/tmp/renovate/repos/gitlab/ad-notam/website\":\"/tmp/renovate/repos/gitlab/ad-notam/website\" -v \"/tmp/renovate/cache\":\"/tmp/renovate/cache\" -e NPM_CONFIG_CACHE -e npm_config_store -e CI -e YARN_CACHE_FOLDER -w \"/tmp/renovate/repos/gitlab/ad-notam/website\" docker.io/renovate/node bash -l -c \"npm i -g yarn && sed -i 's/ steps,/ steps.slice(0,1),/' /home/ubuntu/.npm-global/lib/node_modules/yarn/lib/cli.js && yarn install --ignore-engines --ignore-platform --network-timeout 100000 --ignore-scripts\"\nerror Your lockfile needs to be updated, but yarn was run with `--frozen-lockfile`.\n\n    at ChildProcess.exithandler (child_process.js:319:12)\n    at ChildProcess.emit (events.js:376:20)\n    at ChildProcess.emit (domain.js:470:12)\n    at maybeClose (internal/child_process.js:1055:16)\n    at Process.ChildProcess._handle.onexit (internal/child_process.js:288:5)"
       }
DEBUG: lock file error (repository=ad-notam/website, branch=renovate/webpack-packages)
       "err": {
         "killed": false,
         "code": 1,
         "signal": null,
         "cmd": "docker run --rm --name=renovate_node --label=renovate_child -v \"/tmp/renovate/repos/gitlab/ad-notam/website\":\"/tmp/renovate/repos/gitlab/ad-notam/website\" -v \"/tmp/renovate/cache\":\"/tmp/renovate/cache\" -e NPM_CONFIG_CACHE -e npm_config_store -e CI -e YARN_CACHE_FOLDER -w \"/tmp/renovate/repos/gitlab/ad-notam/website\" docker.io/renovate/node bash -l -c \"npm i -g yarn && sed -i 's/ steps,/ steps.slice(0,1),/' /home/ubuntu/.npm-global/lib/node_modules/yarn/lib/cli.js && yarn install --ignore-engines --ignore-platform --network-timeout 100000 --ignore-scripts\"",
         "stdout": "\n> yarn@1.22.11 preinstall /home/ubuntu/.npm-global/lib/node_modules/yarn\n> :; (node ./preinstall.js > /dev/null 2>&1 || true)\n\n/home/ubuntu/.npm-global/bin/yarn -> /home/ubuntu/.npm-global/lib/node_modules/yarn/bin/yarn.js\n/home/ubuntu/.npm-global/bin/yarnpkg -> /home/ubuntu/.npm-global/lib/node_modules/yarn/bin/yarn.js\n+ yarn@1.22.11\nadded 1 package in 1.013s\nyarn install v1.22.11\n[1/4] Resolving packages...\ninfo Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.\n",
         "stderr": "error Your lockfile needs to be updated, but yarn was run with `--frozen-lockfile`.\n",
         "message": "Command failed: docker run --rm --name=renovate_node --label=renovate_child -v \"/tmp/renovate/repos/gitlab/ad-notam/website\":\"/tmp/renovate/repos/gitlab/ad-notam/website\" -v \"/tmp/renovate/cache\":\"/tmp/renovate/cache\" -e NPM_CONFIG_CACHE -e npm_config_store -e CI -e YARN_CACHE_FOLDER -w \"/tmp/renovate/repos/gitlab/ad-notam/website\" docker.io/renovate/node bash -l -c \"npm i -g yarn && sed -i 's/ steps,/ steps.slice(0,1),/' /home/ubuntu/.npm-global/lib/node_modules/yarn/lib/cli.js && yarn install --ignore-engines --ignore-platform --network-timeout 100000 --ignore-scripts\"\nerror Your lockfile needs to be updated, but yarn was run with `--frozen-lockfile`.\n",
         "stack": "Error: Command failed: docker run --rm --name=renovate_node --label=renovate_child -v \"/tmp/renovate/repos/gitlab/ad-notam/website\":\"/tmp/renovate/repos/gitlab/ad-notam/website\" -v \"/tmp/renovate/cache\":\"/tmp/renovate/cache\" -e NPM_CONFIG_CACHE -e npm_config_store -e CI -e YARN_CACHE_FOLDER -w \"/tmp/renovate/repos/gitlab/ad-notam/website\" docker.io/renovate/node bash -l -c \"npm i -g yarn && sed -i 's/ steps,/ steps.slice(0,1),/' /home/ubuntu/.npm-global/lib/node_modules/yarn/lib/cli.js && yarn install --ignore-engines --ignore-platform --network-timeout 100000 --ignore-scripts\"\nerror Your lockfile needs to be updated, but yarn was run with `--frozen-lockfile`.\n\n    at ChildProcess.exithandler (child_process.js:319:12)\n    at ChildProcess.emit (events.js:376:20)\n    at ChildProcess.emit (domain.js:470:12)\n    at maybeClose (internal/child_process.js:1055:16)\n    at Process.ChildProcess._handle.onexit (internal/child_process.js:288:5)"
       },
       "type": "yarn"
```

</details>

### Have you created a minimal reproduction repository?

I have linked to a minimal reproduction repository in the bug description

### PR DESCRIPTION
**Pull request by aommm**
Haven't contributed to this repo before, please let me know if I'm doing anything wrong! 🙂 

## Changes
Fixes so that Renovate can run on repos that have `--install.frozen-lockfile true` in their .yarnrc. 

## Context
Fixes #11356.

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

## Test log

When running `main` towards my repository, it created the following broken PR ❌:
https://github.com/aommm/renovate_reproduction_11356/pull/2

When running `aommm:fix/npm-yarnrc-frozen-lockfile` towards my repository, it created the following working PR ✅:
https://github.com/aommm/renovate_reproduction_11356/pull/3

Also it logged as I expected:
```
DEBUG: Writing .yarnrc to . (repository=aommm/renovate_reproduction_11356, branch=renovate/webpack-packages)
```

### CODE REVIEW

File: lib/modules/manager/npm/post-update/__fixtures__/yarnrc-massage-1/.yarnrc [hunk 1]
```diff
@@ -0,0 +1,2 @@
+--install.pure-lockfile true
+--install.frozen-lockfile true
```
Comment (rarkins) [inline lib/modules/manager/npm/post-update/__fixtures__/yarnrc-massage-1/.yarnrc:2]: No need for an extra file - please include it as a string in the test

File: lib/modules/manager/npm/post-update/index.ts [hunk 2]
```diff
@@ -241,6 +241,25 @@ export async function writeExistingFiles(
...
+
+    const yarnrcFilename = upath.join(basedir, '.yarnrc');
+    let yarnrcContent: string | null = null;
+    try {
+      yarnrcContent = await getFile(yarnrcFilename);
+    } catch (err) /* istanbul ignore next */ {
+      // .yarnrc doesn't exist
+    }
+    if (yarnrcContent) {
+      logger.debug(`Writing .yarnrc to ${basedir}`);
```
Comment (rarkins) [inline lib/modules/manager/npm/post-update/index.ts:253]: Better if you log and write only if there's changes to the content

**Review by rarkins — CHANGES_REQUESTED** [review]

**Comment by rarkins** [comment]
Thanks for your first contribution! Looking good but some small changes resisted.

Do you have a reproduction repo which today fails?

**Comment by aommm** [comment]
Thanks for the comments, have addressed them now.

I have a reproduction repo [here](https://github.com/aommm/renovate_reproduction_11356/pulls?q=is%3Apr+is%3Aclosed) (forked from @afdev82).
I ran my fix towards it and it created a working PR: https://github.com/aommm/renovate_reproduction_11356/pull/3
For comparison, `main` created a broken PR: https://github.com/aommm/renovate_reproduction_11356/pull/2

So I think that this works and ticks all the boxes now, so it's ready for review again 🙂

**Comment by aommm** [comment]
actually I broke some of the other tests, pushed a fix for that now.

File: lib/modules/manager/npm/post-update/index.ts [hunk 3]
```diff
@@ -241,6 +241,27 @@ export async function writeExistingFiles(
...
+    } catch (err) /* istanbul ignore next */ {
+      // .yarnrc doesn't exist
+    }
+    if (yarnrcContent) {
+      const yarnrcContentNew = yarnrcContent
+        .replace('--install.pure-lockfile true', '')
+        .replace('--install.frozen-lockfile true', '');
+      if (yarnrcContent !== yarnrcContentNew) {
+        try {
+          logger.debug(`Writing .yarnrc to ${basedir}`);
```
Comment (rarkins) [inline lib/modules/manager/npm/post-update/index.ts:258]:
```suggestion
          logger.debug(`Writing massaged .yarnrc to ${basedir}`);
```

File: lib/modules/manager/npm/post-update/index.ts [hunk 4]
```diff
@@ -241,6 +241,27 @@ export async function writeExistingFiles(
     if (packageFile.pnpmShrinkwrap && config.reuseLockFiles === false) {
       await deleteLocalFile(packageFile.pnpmShrinkwrap);
     }
+
+    const yarnrcFilename = upath.join(basedir, '.yarnrc');
```
Comment (rarkins) [inline lib/modules/manager/npm/post-update/index.ts:245]: Can't this logic be in a yarn-specific section

**Review by rarkins — CHANGES_REQUESTED** [review]

File: lib/modules/manager/npm/post-update/index.ts (hunk 4 above)
Comment (aommm) [reply lib/modules/manager/npm/post-update/index.ts:245]: That's a good idea. Since I don't know the codebase that well I would appreciate some assistance in this. Where do you think is appropriate?

Here are some alternatives I can think of:

- New function `massageYarnrc` within `index.ts`
- New function `massageYarnrc` in `yarn.ts`
- Modify `checkYarnrc` in `yarn.ts`

Comment (rarkins) [reply lib/modules/manager/npm/post-update/index.ts:245]: I think the latter because you can see it already writes out a modified yarnrc here: https://github.com/renovatebot/renovate/blob/4d8e2c704f15b37140f671601cf361c931d28421/lib/modules/manager/npm/post-update/yarn.ts#L64-L67

Comment (aommm) [reply lib/modules/manager/npm/post-update/index.ts]: Have fixed this now, can you take a look again?

I also ran it against my repro repo again and it seems to work:
https://github.com/aommm/renovate_reproduction_11356/pull/4

**Review by rarkins — APPROVED** [review]

**Comment by aommm** [comment]
@rarkins I think this is ready to be merged now.

I'm not sure what strategy you follow regarding merging & versioning. Maybe you are holding off deliberately, which is fine for me.

But I thought I'd mention it in case you had missed to merge this.

File: lib/modules/manager/npm/post-update/yarn.ts [hunk 5]
```diff
@@ -56,16 +56,22 @@ export async function checkYarnrc(
...
         );
+        yarnPath = null;
+      }
+      if (yarnrc !== scrubbedYarnrc) {
+        logger.debug(`Writing scrubbed .yarnrc to ${lockFileDir}`);
         await writeLocalFile(
           upath.join(lockFileDir, '.yarnrc'),
           scrubbedYarnrc
         );
-        yarnPath = null;
```
Comment (viceice) [inline lib/modules/manager/npm/post-update/yarn.ts:68]: why this change?

Comment (aommm) [reply lib/modules/manager/npm/post-update/yarn.ts:68]: It is moved to above. I wanted to preserve its location at the end of the `if (!yarnBinaryExists)` block, because it felt related to that.

Happy to change it if you think that that makes more sense though.

Comment (viceice) [reply lib/modules/manager/npm/post-update/yarn.ts:68]: no, it's just sometimes hard to follow changes. leave as is

**Review by viceice — APPROVED** [review]
